server/schedule/operator/step.go

```diff
@@ -438,3 +440,331 @@ func (al AddLightLearner) Influence(opInfluence OpInfluence, region *core.Region
+		}
+		return peer.GetId() == dv.PeerID
+	}
+	return false
+}
+
+// ChangePeerV2Enter is an OpStep that uses joint consensus to request all PromoteLearner and DemoteVoter.
+type ChangePeerV2Enter struct {
+	PromoteLearners []PromoteLearner
+	DemoteVoters    []DemoteVoter
+}
+
+func (cpe ChangePeerV2Enter) String() string {
+	b := &strings.Builder{}
+	_, _ = b.WriteString("use joint consensus")
+	for _, pl := range cpe.PromoteLearners {
+		_, _ = fmt.Fprintf(b, ", promote learner peer %v on store %v to voter", pl.PeerID, pl.ToStore)
+	}
+	for _, dv := range cpe.DemoteVoters {
+		_, _ = fmt.Fprintf(b, ", demote voter peer %v on store %v to learner", dv.PeerID, dv.ToStore)
+	}
+	return b.String()
+}
+
+// ConfVerChanged returns the delta value for version increased by this step.
+func (cpe ChangePeerV2Enter) ConfVerChanged(region *core.RegionInfo) uint64 {
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStoreVoter(pl.ToStore)
+		if peer.GetId() != pl.PeerID || !core.ReadyToBecomeVoter(peer) {
+			return 0
+		}
+	}
+	for _, dv := range cpe.DemoteVoters {
+		peer := region.GetStoreVoter(dv.ToStore)
+		if peer != nil && (peer.GetId() != dv.PeerID || !core.ReadyToBecomeLearner(peer)) {
+			return 0
+		}
+	}
+	return uint64(len(cpe.PromoteLearners) + len(cpe.DemoteVoters))
+}
+
+// IsFinish checks if current step is finished.
+func (cpe ChangePeerV2Enter) IsFinish(region *core.RegionInfo) bool {
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStoreVoter(pl.ToStore)
+		if peer != nil && peer.GetId() != pl.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", pl.String()), zap.Uint64("obtain-voter", peer.GetId()))
+		}
+		if peer.GetId() != pl.PeerID || peer.GetRole() != metapb.PeerRole_IncomingVoter {
+			return false
+		}
+	}
+	for _, dv := range cpe.DemoteVoters {
+		peer := region.GetStoreVoter(dv.ToStore)
+		if peer != nil && peer.GetId() != dv.PeerID {
+			log.Warn("obtain unexpected peer", zap.String("expect", dv.String()), zap.Uint64("obtain-learner", peer.GetId()))
+		}
+		if peer.GetId() != dv.PeerID || peer.GetRole() != metapb.PeerRole_DemotingVoter {
+			return false
+		}
+	}
+	return true
+}
+
+// CheckSafety checks if the step meets the safety properties.
+func (cpe ChangePeerV2Enter) CheckSafety(region *core.RegionInfo) error {
+	inJointState, notInJointState := false, false
+	for _, pl := range cpe.PromoteLearners {
+		peer := region.GetStorePeer(pl.ToStore)
+		if peer.GetId() != pl.PeerID {
+			return errors.New("peer does not exist")
+		}
+		switch peer.GetRole() {
+		case metapb.PeerRole_Learner:
+			notInJointState = true
+		case metapb.PeerRole_IncomingVoter:
+			inJointState = true
+		case metapb.PeerRole_Voter:
+			return errors.New("peer already is a voter")
+		case metapb.PeerRole_DemotingVoter:
+			return errors.New("cannot promote a demoting voter")
+		default:
+			return errors.New("unexpected peer role")
```

Comment:
What about adding role in error info to help positioning error?

Reply:
In the place where the error log is printed, all peers information will be printed out.

Reply:
got it.